### PR TITLE
refactoring: refactored the FilterState object field support

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: refactoring: refactored the FilterState object field support
Additional Description:

The filter state reflection provides a great feature to access the inner status/property of filter state. However, it has two limitations:
1. It requires the object key be same with the factory key. This limitation make we cannot set multiple objects that with same type.
2. It is a little complex to enable the Field support. We need to define additional reflection class and a factory class.

This PR make things much simpler.

Risk Level: low.
Testing: n/a.
Docs Changes: n/a.
Release Notes: n/a.
Platform Specific Features: n/a.

### CODE REVIEW

File: source/extensions/filters/common/expr/context.cc [hunk 1]
```diff
@@ -339,17 +338,11 @@ absl::optional<CelValue> FilterStateWrapper::operator[](CelValue key) const {
...
-          auto* raw_reflection = reflection.release();
-          arena_.Own(raw_reflection);
-          return CelValue::CreateMap(
-              ProtobufWkt::Arena::Create<FilterStateObjectWrapper>(&arena_, raw_reflection));
-        }
+      // TODO(wbpcode): the implementation of cannot handle the case where the object has provided
+      // field support, but callers only want to access the whole object.
+      if (object->hasFieldSupport()) {
+        return CelValue::CreateMap(
+            ProtobufWkt::Arena::Create<FilterStateObjectWrapper>(&arena_, object));
```
Comment (wbpcode) [inline source/extensions/filters/common/expr/context.cc:345]: I think this is a little weird. We always treat the filter state object that has reflection/field support as map. So, we cannot support the case where the object has related support but we want's to access the whole object (by serializeAsString()).

File: envoy/stream_info/filter_state.h [hunk 2]
```diff
@@ -102,21 +104,17 @@ class FilterState {
...
-  class ObjectReflection {
-  public:
-    virtual ~ObjectReflection() = default;
 
-    using FieldType = absl::variant<absl::monostate, absl::string_view, int64_t>;
+    /**
+     * @return bool true if the object supports field access. False if the object does not support
+     * field access. Default implementation returns false.
+     */
+    virtual bool hasFieldSupport() const { return false; }
```
Comment (wbpcode) [inline envoy/stream_info/filter_state.h:112]: This is only used to support the case where we need to provide different behavior for same object and same calling. (Now, only the expr need this. If we can change the expr in future, we can remove this method.)

**Comment by kyessenov** [comment]
I think we can go ahead with this if we don't anticipate nested objects in the reflected values. The idea was to add `ObjectReflection` as another field type, similar to how it's done in Scala/Swift (https://bracha.org/mirrors.pdf), but we don't need that complexity.

**Comment by wbpcode** [comment]
> I think we can go ahead with this if we don't anticipate nested objects in the reflected values. The idea was to add `ObjectReflection` as another field type, similar to how it's done in Scala/Swift (https://bracha.org/mirrors.pdf), but we don't need that complexity.


Yeah. It's pretty great as a common reflection support for type system. But FilterState::Object is an domain-specific abstraction and needn't that complexity. 

A single new virtual method is much more straight-forward and easier to use for developers.

File: source/extensions/filters/common/expr/context.cc (hunk 1 above)
Comment (kyessenov) [reply source/extensions/filters/common/expr/context.cc:345]: Yes, I'm aware of this limitation. It's same with proto Struct.

**Review by kyessenov — APPROVED** [review]

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**